rust/arrow/src/compute/kernels/trigonometry.rs

```diff
@@ -0,0 +1,246 @@
+pub fn sin_cos<T>(array: &PrimitiveArray<T>) -> (PrimitiveArray<T>, PrimitiveArray<T>)
+where
+    T: ArrowNumericType,
+    T::Native: num::traits::Float,
+{
+    let (sin, cos): (Vec<T::Native>, Vec<T::Native>) =
+        array.values().iter().map(|v| v.sin_cos()).unzip();
+    // NOTE: due to `unzip` collecting and splitting the arrays,
+    // this might be slower than Buffer::from_trusted_len_iter
+    let sin_buffer = Buffer::from_iter(sin);
+    let cos_buffer = Buffer::from_iter(cos);
+
+    let sin_data = into_primitive_array_data::<_, T>(array, sin_buffer);
+    let cos_data = into_primitive_array_data::<_, T>(array, cos_buffer);
+    (
+        PrimitiveArray::<T>::from(std::sync::Arc::new(sin_data)),
+        PrimitiveArray::<T>::from(std::sync::Arc::new(cos_data)),
+    )
+}
+
+/// Calculate the Haversine distance between two geographic coordinates.
+/// Based on the haversine crate.
+///
+/// The distance returned is in meters, and the radius must be specified in meters.
+pub fn haversine<T>(
+    lat_a: &PrimitiveArray<T>,
+    lng_a: &PrimitiveArray<T>,
+    lat_b: &PrimitiveArray<T>,
+    lng_b: &PrimitiveArray<T>,
+    radius: impl num::traits::Float,
+) -> Result<PrimitiveArray<T>>
+where
+    T: ArrowPrimitiveType + ArrowNumericType,
+    T::Native: num::traits::Float + Div<Output = T::Native> + Zero + num::NumCast,
+{
+    // Check array lengths, must all equal
+    let len = lat_a.len();
+    if lat_b.len() != len || lng_a.len() != len || lng_b.len() != len {
+        return Err(ArrowError::ComputeError(
+            "Cannot perform math operation on arrays of different length".to_string(),
+        ));
+    }
+    // These casts normally get optimized to f64 as f64
+    let one = num::cast::<_, T::Native>(1.0).unwrap();
+    let two = num::cast::<_, T::Native>(2.0).unwrap();
+    let radius = num::cast::<_, T::Native>(radius).unwrap();
+    let two_radius = two * radius;
+
+    let lat_delta = to_radians(&subtract(lat_b, lat_a)?);
+    let lng_delta = to_radians(&subtract(lng_b, lng_a)?);
+    let lat_a_rad = to_radians(lat_a);
+    let lat_b_rad = to_radians(lat_b);
+
+    let v1: PrimitiveArray<T> = sin(&unary::<_, _, _>(&lat_delta, |x| x / two));
+    let v2: PrimitiveArray<T> = sin(&unary::<_, _, _>(&lng_delta, |x| x / two));
+
+    let a = add(
+        // powf is slower than x * x
+        &unary::<_, _, _>(&v1, |x: T::Native| x * x),
+        // This could be simplified if we had a ternary kernel that takes 3 args
+        // F(T::Native, T::Native, T::Native) -> T::Native
+        &multiply(
+            // powf is slower than x * x
+            &unary::<_, _, _>(&v2, |x: T::Native| x * x),
+            &math_op(&lat_a_rad, &lat_b_rad, |x, y| (x.cos() * y.cos()))?,
+        )?,
+    )?;
+    Ok(unary::<_, _, _>(
+        &atan2(&sqrt(&a), &sqrt(&unary::<_, _, _>(&a, |x| one - x)))?,
+        |x| x * two_radius,
+    ))
+}
+
+// only added here to test that it's accurate
+pub fn haversine_no_unary<T>(
```

Comment:
shouldn't we move this to the tests, if it is only used for testing?

Reply:
I've removed it, as it's also in the benchmark